File: src/data/courses/dsa/two-pointers-sliding-window/two-pointers-opposite-direction/trapping-rain-water/code/java/TrappingRainWaterSol3.java
class Solution {
    public int trap(int[] h) {
        int l=0, r=h.length-1, ml=0, mr=0, w=0;
        while (l < r) {
            if (h[l] < h[r]) { if (h[l]>=ml) ml=h[l]; else w+=ml-h[l]; l++; }
            else { if (h[r]>=mr) mr=h[r]; else w+=mr-h[r]; r--; }
        }
        return w;
    }
}

File: src/data/courses/dsa/two-pointers-sliding-window/two-pointers-opposite-direction/trapping-rain-water/code/python/trapping_rain_water_sol3.py
class Solution:
    def trap(self, height):
        l, r = 0, len(height)-1
        max_l = max_r = water = 0
        while l < r:
            if height[l] < height[r]:
                if height[l] >= max_l: max_l = height[l]
                else: water += max_l - height[l]
                l += 1
            else:
                if height[r] >= max_r: max_r = height[r]
                else: water += max_r - height[r]
                r -= 1
        return water